middlewares/retry.go

```diff
@@ -32,12 +33,19 @@ func NewRetry(attempts int, next http.Handler) *Retry {
 }
 
 func (retry *Retry) ServeHTTP(rw http.ResponseWriter, r *http.Request) {
+	// if we might make multiple attempts, swap the body for an ioutil.NopCloser
+	// cf https://github.com/containous/traefik/issues/1008
+	if retry.attempts > 1 {
+		body := r.Body
+		defer body.Close()
+		r.Body = ioutil.NopCloser(body)
+	}
 	attempts := 1
 	for {
 		recorder := NewRecorder()
 		recorder.responseWriter = rw
 		retry.next.ServeHTTP(recorder, r)
-		if !isNetworkError(recorder.Code) || attempts >= retry.attempts {
+		if recorder.Error == nil || attempts >= retry.attempts {
```

Comment:
Would you consider these as network errors too? At least from my pov, if the forwarder doesn't manage to send a request or get a proper response from upstream for whatever reason I'd see it as a network error, as opposed to a valid http response from upstream with a 5XX status code, which would currently be retried.

Shall I stick to the first commit and open an other issue to discuss about the exact behaviour we want for the retry feature?